pkg/flow/internal/controller/node_config_import.go

```diff
@@ -0,0 +1,459 @@
+		}
+	}
+}
+
+// processDeclareBlock processes a declare block.
+func (cn *ImportConfigNode) processDeclareBlock(stmt *ast.BlockStmt, content string) {
+	cn.contentMut.Lock()
+	defer cn.contentMut.Unlock()
+	if _, ok := cn.importedDeclares[stmt.Label]; ok {
+		level.Error(cn.logger).Log("msg", "declare block redefined", "name", stmt.Label)
+		return
+	}
+	cn.importedDeclares[stmt.Label] = NewDeclare(stmt, content[stmt.LCurlyPos.Position().Offset+1:stmt.RCurlyPos.Position().Offset-1])
+}
+
+// processDeclareBlock processes an import block.
+func (cn *ImportConfigNode) processImportBlock(stmt *ast.BlockStmt, fullName string) {
+	sourceType := importsource.GetSourceType(fullName)
+	if _, ok := cn.importConfigNodesChildren[stmt.Label]; ok {
+		level.Error(cn.logger).Log("msg", "import block redefined", "name", stmt.Label)
+		return
+	}
+	childGlobals := cn.globals
+	// Children have a special OnNodeWithDependantsUpdate function which will surface all the imported declares to the root import config node.
+	childGlobals.OnNodeWithDependantsUpdate = cn.OnChildrenContentUpdate
+	cn.importConfigNodesChildren[stmt.Label] = NewImportConfigNode(stmt, childGlobals, sourceType)
+}
+
+// onContentUpdate is triggered every time the managed import component has new content.
+func (cn *ImportConfigNode) onContentUpdate(content string) {
+	cn.importChildrenMut.Lock()
+	defer cn.importChildrenMut.Unlock()
+	cn.importConfigNodesChildren = make(map[string]*ImportConfigNode)
+
+	cn.contentMut.Lock()
+	cn.inContentUpdate = true
+	defer func() {
+		cn.contentMut.Lock()
+		cn.inContentUpdate = false
+		cn.contentMut.Unlock()
+	}()
+	cn.importedDeclares = make(map[string]*Declare)
+	cn.contentMut.Unlock()
+
+	node, err := parser.ParseFile(cn.label, []byte(content))
+	if err != nil {
+		level.Error(cn.logger).Log("msg", "failed to parse file on update", "err", err)
+		return
+	}
+	cn.processNodeBody(node, content)
+	err = cn.evaluateChildren()
+	if err != nil {
+		level.Error(cn.logger).Log("msg", "failed to evaluate nested import", "err", err)
+		return
+	}
+
+	if cn.importChildrenRunning {
+		cn.importChildrenUpdateChan <- struct{}{}
+	}
+
+	cn.lastUpdateTime.Store(time.Now())
+	cn.OnNodeWithDependantsUpdate(cn)
+}
+
+// evaluateChildren evaluates the import nodes managed by this import node.
+func (cn *ImportConfigNode) evaluateChildren() error {
+	for _, child := range cn.importConfigNodesChildren {
+		err := child.Evaluate(&vm.Scope{
+			Parent:    nil,
+			Variables: make(map[string]interface{}),
+		})
+		if err != nil {
+			return fmt.Errorf("imported node %s failed to evaluate, %v", child.label, err)
+		}
+	}
+	return nil
+}
+
+// runChildren run the import nodes managed by this import node.
+// The children list can be updated onContentUpdate. In this case we need to stop the running children and run the new set of children.
+func (cn *ImportConfigNode) runChildren(parentCtx context.Context) error {
```

Comment:
This function is not trivial, I added a check for leaky goroutines and it appears that every few test that one go routine inside of start children is leaky. Not sure whether it is a test error or if there is a real leak. I'd be happy to get feedback on how this area can be improved